src/common/dateInput/DateInputLogic.tsx

```diff
@@ -0,0 +1,240 @@
+    .join('');
+}
+
+function cutToLength(string: string, length: number) {
+  if (string.length <= length) {
+    return string;
+  }
+
+  return string.slice(0, length);
+}
+
+function parseNumberFromValue(value: string, maxCharacterCount = 2): number {
+  // Removes all other characters than numbers from the result. This is
+  // a last fallback which makes negative numbers positive and removes
+  // possible characters from the output. The behavior of
+  // input[type="number"] is not consistent between platforms.
+  const sanitized = sanitizeInput(value);
+  const maxSized = cutToLength(sanitized, maxCharacterCount);
+
+  return Number(maxSized);
+}
+
+function DateInputLogic({
+  onChange,
+  value,
+  isInvalid,
+  wrapper,
+  dateInputId,
+  monthInputId,
+  yearInputId,
+  dateInputName,
+  monthInputName,
+  yearInputName,
+  dateComponent,
+  monthComponent,
+  yearComponent,
+  dateInputLabel,
+  monthInputLabel,
+  yearInputLabel,
+}: Props) {
+  const currentDate = getDateComponents(value);
+  const [cachedDate, setCachedDate] = React.useState<DateObject>(currentDate);
+  const monthInputRef = React.useRef(null);
+  const yearInputRef = React.useRef(null);
+
+  const updateDate = (
+    date: number | null,
+    month: number | null,
+    year: number | null
+  ) => {
+    setCachedDate(previousDate => {
+      const nextCachedDate = {
+        date: defaultTo(date, previousDate.date),
+        month: defaultTo(month, previousDate.month),
+        year: defaultTo(year, previousDate.year),
+      };
+      const nextDate = makeDate(nextCachedDate);
+
+      if (nextDate) {
+        onChange(nextDate);
+      }
+
+      return nextCachedDate;
+    });
+  };
+
+  const focusWhen = (
+    condition: boolean,
+    element: React.RefObject<HTMLInputElement>
+  ) => {
+    if (condition && element.current) {
+      element.current.focus();
+    }
+  };
+
+  const handleDateChange = (event: React.FormEvent<HTMLInputElement>) => {
+    const nextDate = parseNumberFromValue(event.currentTarget.value);
+
+    updateDate(nextDate, null, null);
+    focusWhen(getCharacterCount(nextDate) === 2, monthInputRef);
+  };
+  const handleMonthChange = (event: React.FormEvent<HTMLInputElement>) => {
+    const nextMonth = parseNumberFromValue(event.currentTarget.value);
+    // User's input month number, we save month index into state
+    const nextMonthIndex = nextMonth - 1;
+
+    updateDate(null, nextMonthIndex, null);
+    focusWhen(getCharacterCount(nextMonth) === 2, yearInputRef);
+  };
+  const handleYearChange = (event: React.FormEvent<HTMLInputElement>) => {
+    const nextYear = parseNumberFromValue(event.currentTarget.value, 4);
+
+    updateDate(null, null, nextYear);
+  };
+
+  const formattedMonth =
+    cachedDate.month !== undefined
+      ? (cachedDate.month + 1).toString()
+      : undefined;
+
+  return React.createElement(wrapper, { isInvalid }, [
+    React.createElement(dateComponent, {
```

Comment:
If this component is actually about the day of the month, I think dayComponent would be a better name, so that it is easier to separate it from a "full/normal date"